### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.040251970291137695 seconds
34.4 ms ± 584 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



### Resultados en cola `hpc-bio-mendel` (SLURM)

**Comandos ejecutados:**
```bash
sbatch submit_Python_hpc-bio-mendel-alumno05.sh 10000000
sbatch submit_Python_hpc-bio-mendel-alumno05.sh 100000000

Módulo cargado: anaconda/2025.06
Cola: hpc-bio-mendel
Logs: slurm-lab-python-job-12692.out, slurm-lab-python-job-12693.out
Tiempos y suma obtenida:

N = 10^7

[Python] N=10000000 sum=49999995000000 time=0.545447 s
[Numba ] N=10000000 sum=49999995000000 time=0.000003 s
%timeit: 520 ms ± 0.755 ms (2 runs)


N = 10^8

[Python] N=100000000 sum=4999999950000000 time=5.526836 s
[Numba ] N=100000000 sum=4999999950000000 time=0.000019 s
%timeit: 5.35 s ± 0.00103 s (2 runs)

In [2]:
import time

value = 10**6
lista = list(range(value))

# lista + for
start = time.time()
suma_lista_for = 0
for i in lista:
    suma_lista_for += i
end = time.time()
print("Tiempo (lista + for):", end - start, "s")

# lista + sum
start = time.time()
suma_lista_sum = sum(lista)
end = time.time()
print("Tiempo (lista + sum):", end - start, "s")

Tiempo (lista + for): 0.06048417091369629 s
Tiempo (lista + sum): 0.006242513656616211 s


In [3]:
import numpy as np

array = np.arange(value, dtype=np.int64)

# array + for
start = time.time()
suma_array_for = 0
for i in array:
    suma_array_for += i
end = time.time()
print("Tiempo (array + for):", end - start, "s")

# array + np.sum
start = time.time()
suma_array_sum = np.sum(array)
end = time.time()
print("Tiempo (array + np.sum):", end - start, "s")

Tiempo (array + for): 0.12106156349182129 s
Tiempo (array + np.sum): 0.0004582405090332031 s


In [4]:
from numba import njit

@njit
def reduc_numba(a):
    x = 0
    for i in range(a):
        x += i
    return x

# warm-up
_ = reduc_numba(10)

start = time.time()
suma_numba = reduc_numba(value)
end = time.time()
print("Tiempo (numba njit):", end - start, "s")

Tiempo (numba njit): 5.221366882324219e-05 s


Comparación de métodos (N=10^6)

- Bucle Python → 0.040 s  
- Lista + for → 0.060 s  
- Lista + sum → 0.0062 s  
- Array + for → 0.121 s  
- Array + np.sum → 0.00046 s  
- Numba (njit) → 0.000052 s  

Conclusión: sum y np.sum son mucho más rápidos que el bucle normal, y Numba es el que más acelera.